src/DynamicData/Binding/IObservableListEx.cs

```diff
@@ -0,0 +1,197 @@
+        /// <typeparam name="TKey">The type of the key.</typeparam>
+        /// <param name="change">The source change</param>
+        /// <param name="list">
+        /// An optional list, if provided it allows the refresh from a key based cache to find the index for the resulting list based refresh.
+        /// If not provided a refresh will dropdown to a replace which may ultimately result in a remove+add change downstream.
+        /// </param>
+        /// <returns>The downcasted <see cref="Change{TObject}"/></returns>
+        private static Change<TObject> Convert<TObject, TKey>(Change<TObject, TKey> change, IExtendedList<TObject> list = null)
+        {
+            switch (change.Reason)
+            {
+                case ChangeReason.Add:
+                    // If not sorted
+                    if (change.CurrentIndex == -1)
+                    {
+                        return new Change<TObject>(ListChangeReason.Add, change.Current);
+                    }
+
+                    return new Change<TObject>(ListChangeReason.Add, change.Current, index: change.CurrentIndex);
+
+                case ChangeReason.Moved:
+                    // Move is always sorted
+                    return new Change<TObject>(change.Current, change.CurrentIndex, change.PreviousIndex);
+                case ChangeReason.Refresh:
+                    // Refresh needs an index, which we don't have in a Change<T, K> model since it's key based.
+                    // See: DynamicData > Binding > ObservableCollectionAdaptor.cs Line 129-130
+
+                    // Note: A refresh is not index based within the context of a sorted change.
+                    // Thus, currentIndex will not be available here where as other changes like add and remove do have indexes if coming from a sorted changeset.
+
+                    // In order to properly handle a refresh and map to an index on a list, we need to use the source list (within the edit method so that it's thread safe)
+                    if (list != null && list.IndexOf(change.Current) is int index && index >= 0)
+                    {
+                        return new Change<TObject>(ListChangeReason.Refresh, current: change.Current, index: index);
+                    }
+
+                    // Fallback to a replace if a list is not available
+                    return new Change<TObject>(ListChangeReason.Replace, current: change.Current, previous: change.Current);
+                case ChangeReason.Remove:
+                    // If not sorted
+                    if (change.CurrentIndex == -1)
+                    {
+                        return new Change<TObject>(ListChangeReason.Remove, change.Current);
+                    }
+                    return new Change<TObject>(ListChangeReason.Remove, change.Current, index: change.CurrentIndex);
+
+                case ChangeReason.Update:
+                    // If not sorted
```

Comment:
Semantically, I always intended a replace to be the list equivalent of `list[index]=value` rather than the direct equivalent `dict[key]=value`. I am not sure whether treating these the same for all down stream operators will produce consistent results.

Reply:
I see - when adapting `RemoveKeyEnumerator<TObject, TKey>` we left it as a remove/add. Let us know if there is any adjustments we still need to make. Thanks!